kedro/io/kedro_data_catalog.py

```diff
@@ -0,0 +1,345 @@
+        """
+        catalog = catalog or {}
+        config_resolver = CatalogConfigResolver(catalog, credentials)
+        save_version = save_version or generate_timestamp()
+        load_versions = load_versions or {}
+
+        missing_keys = [
+            ds_name
+            for ds_name in load_versions
+            if not (
+                ds_name in config_resolver.config
+                or config_resolver.match_pattern(ds_name)
+            )
+        ]
+        if missing_keys:
+            raise DatasetNotFoundError(
+                f"'load_versions' keys [{', '.join(sorted(missing_keys))}] "
+                f"are not found in the catalog."
+            )
+
+        return cls(
+            load_versions=load_versions,
+            save_version=save_version,
+            config_resolver=config_resolver,
+        )
+
+    @staticmethod
+    def _validate_dataset_config(ds_name: str, ds_config: Any) -> None:
+        if not isinstance(ds_config, dict):
+            raise DatasetError(
+                f"Catalog entry '{ds_name}' is not a valid dataset configuration. "
+                "\nHint: If this catalog entry is intended for variable interpolation, "
+                "make sure that the key is preceded by an underscore."
+            )
+
+    def _add_from_config(self, ds_name: str, ds_config: dict[str, Any]) -> None:
+        # TODO: Add lazy loading feature to store the configuration but not to init actual dataset
+        # TODO: Initialise actual dataset when load or save
+        self._validate_dataset_config(ds_name, ds_config)
+        ds = AbstractDataset.from_config(
+            ds_name,
+            ds_config,
+            self._load_versions.get(ds_name),
+            self._save_version,
+        )
+
+        self.add(ds_name, ds)
+
+    def get_dataset(
+        self, ds_name: str, version: Version | None = None, suggest: bool = True
+    ) -> AbstractDataset:
+        """Get a dataset by name from an internal collection of datasets.
+
+        If a dataset is not in the collection but matches any pattern
+        it is instantiated and added to the collection first, then returned.
+
+        Args:
+            ds_name: A dataset name.
+            version: Optional argument for concrete dataset version to be loaded.
+                Works only with versioned datasets.
+            suggest: Optional argument whether to suggest fuzzy-matching datasets' names
+                in the DatasetNotFoundError message.
+
+        Returns:
+            An instance of AbstractDataset.
+
+        Raises:
+            DatasetNotFoundError: When a dataset with the given name
+                is not in the collection and do not match patterns.
+        """
+        ds_config = self._config_resolver.resolve_dataset_pattern(ds_name)
+
+        if ds_name not in self._datasets and ds_config:
+            self._add_from_config(ds_name, ds_config)
+
+        dataset = self._datasets.get(ds_name, None)
+
+        if dataset is None:
```

Comment:
Well, we can only fail after we try to resolve. Otherwise, you get one more layer of `if` as the logic needs to go inside the `if fail [] else []` scenario.

Now the logic is like this:

- if not part of the materialised datasets, resolve the dataset pattern
- if resolved, add from config
- get the dataset
- if the dataset does not exist (basically if it cannot be resolved nor exists), go with the error scenario
- otherwise, continue with a non-error scenario

Reply:
What I meant is that we should have as a first line something like:

```
if ds_name not in self._datasets and self._config_resolver.match_pattern():
    ...
```

and then continue with the error scenario, and then go on with everything else. It'll be much easier to follow this way. Btw while checking if this is possible, I saw a problem in the resolver - it can fail even if it matches, but that should not happen. 

https://github.com/kedro-org/kedro/blob/5147dfb24d8f1468daadacb8140406cb13485e8a/kedro/io/catalog_config_resolver.py#L149-L156

This ☝️ should have been checked at the config_resolver init time, basically we should not allow to create a config_resolver with unresolvable configs or add invalid configs that cannot be resolved.

Also there's other changes like e.g. `resolve_dataset_pattern` should be just `resolve_pattern` similar to all other public methods there, which never include the word `dataset` (rightfully). Hopefully the resolver API is not released yet, so we can change it now.

Reply:
1. Checking for a pattern match is not enough, as the config could already be resolved at the init time. `resolve_pattern` method encapsulates this, not to expose this logic outside the `config_resolver`. So we ask the `config_resolver` to provide a config for a pattern without bothering about how it's happening inside. I don't think we need to move any resolution logic (including any specific checks) to the catalog level.
2. The configs resolution makes sense to me. We can move this validation to the init time and simplify the resolution method. But would do that in a separate PR as it doesn't touch the catalog and will be done at the level of the config resolver.
3. Method was renamed.

Reply:
1. Why wouldn't it be enough? We are checking only for failing scenarios, what other failing scenarios would there be apart from not matching a concrete dataset or a pattern? Could we expect failure of instantiation?
2. ✅ 
3. ✅ 

In any case, this is a minor thing, let's merge it in as it is and then we can always come back and simplify it.